Assets/MRTK/SDK/Features/Utilities/Solvers/HandConstraintPalmUp.cs

```diff
@@ -163,5 +197,125 @@ protected override bool IsValidController(IMixedRealityController controller)
+
+                if (jointedHand.TryGetJoint(TrackedHandJoint.IndexTip, out indexTipPose) &&
+                    jointedHand.TryGetJoint(TrackedHandJoint.RingTip, out ringTipPose))
+                {
+                    var handNormal = Vector3.Cross(indexTipPose.Position - palmPose.Position,
+                                                   ringTipPose.Position - indexTipPose.Position).normalized;
+                    handNormal *= (jointedHand.ControllerHandedness == Handedness.Right) ? 1.0f : -1.0f;
+
+                    if (Vector3.Angle(palmPose.Up, handNormal) > flatHandThreshold)
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            // Check if the palm angle meets the prescribed threshold
+            return palmCameraAngle < facingCameraTrackingThreshold;
+        }
+
+        /// <summary>
+        /// Checks to see if the user is currently gazing at the activation point; it first attempts to do so 
+        /// using eyegaze, and then falls back to head-based gaze if eyegaze isn't available for use.
+        /// </summary>
+        /// <param name="jointedHand"></param>
+        /// <returns></returns>
+        private bool IsUserGazingAtActivationPoint(IMixedRealityHand jointedHand)
+        {
+            Ray gazeRay;
+
+            if (InputRayUtils.TryGetRay(InputSourceType.Eyes, Handedness.Any, out gazeRay) || InputRayUtils.TryGetRay(InputSourceType.Head, Handedness.Any, out gazeRay))
+            {
+                // Generate the hand plane that we're using to generate a distance value.
+                // This is done by using the index knuckle, pinky knuckle, and wrist
+                MixedRealityPose indexKnuckle;
+                MixedRealityPose pinkyKnuckle;
+                MixedRealityPose wrist;
+
+                if (jointedHand.TryGetJoint(TrackedHandJoint.IndexKnuckle, out indexKnuckle) &&
+                    jointedHand.TryGetJoint(TrackedHandJoint.PinkyKnuckle, out pinkyKnuckle) &&
+                    jointedHand.TryGetJoint(TrackedHandJoint.Wrist, out wrist))
+                {
+                    Plane handPlane = new Plane(indexKnuckle.Position, pinkyKnuckle.Position, wrist.Position);
+                    float distanceToHandPlane;
+
+                    if (handPlane.Raycast(gazeRay, out distanceToHandPlane))
+                    {
+                        // Define the activation point as a vector between the wrist and pinky knuckle; then cast it against the plane to get a smooth location
+                        Vector3 activationPoint = Vector3.Lerp(pinkyKnuckle.Position, wrist.Position, .5f);
+
+                        // Now that we know the dist to the plane, create a vector at that point
+                        Vector3 gazePosOnPlane = gazeRay.origin + gazeRay.direction.normalized * distanceToHandPlane;
+                        Vector3 PlanePos = handPlane.ClosestPointOnPlane(gazePosOnPlane);
```

Comment:
Have you profiled the cost of doing this every frame? I suspect it adds a considerable overhead.

Reply:
I haven't compared it quite yet, but so far there's been no indication (on-device) that it causes any performance hit. I'll run a profile and post the results here for visibility.